### PR TITLE
IGNITE-14658 Adds connector SSL metrics.

### PR DESCRIPTION
**Pull request by ololo3000**
The PR adds the following new metrics:

sslHandshakeDurationHistogram  - provides the histogram of SSL handshake duration c.
sslRejectedSessionsCount - provides the number of rejected connections due to SSL errors (not exported if ssl is disabled).
sslEnabled - indicates whether SSL is enabled for the connector
activeSessionsCount - provides the number of active TCP sessions

Each of the mentioned above metrics was added to the following metric groups that relate to different connectors (SSL is configured separately for each of them)

(NEW) client.connector.tcp - for metrics related to thin client connector SSL metrics.
(NEW) client.connector.rest.tcp - for metrics related to REST TCP client connector SSL metrics.
communication.tcp - for metrics related to node communicatonconnector SSL metrics.

The following metrics are also added to the discovery metric group - 
io.discovery.sslRejectedConnectionsMetric - tracks number of join connections that were rejected due to SSL errors ((not exported if ssl is disabled))
io.discovery.sslEnabled - tracks whether SSL is enabled for the discovery

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/protocols/tcp/GridTcpRestProtocol.java [hunk 1]
```diff
@@ -63,6 +63,9 @@
     /** NIO server listener. */
     private GridTcpRestNioListener lsnr;
 
+    /** The name of the metric group associated with the REST TCP connector. */
+    public static final String REST_CONNECTOR_METRIC_GROUP_NAME = "client.connector.rest.tcp";
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/protocols/tcp/GridTcpRestProtocol.java:67]: please, use `MetricUtils.metricName` here.

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/protocols/tcp/GridTcpRestProtocol.java:67]: `GROUP` -> `REGISTRY`

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    /** Tests SSL metrics produced by JDBC connection. */
+    @Test
+    public void testJdbc() throws Exception {
+        MetricRegistry reg = mreg(startClusterNode(0), CLIENT_CONNECTOR_METRIC_REGISTRY_NAME);
+
+        try (Connection ignored = getConnection(jdbcConfiguration("thinClient", "trusttwo", CIPHER_SUITE, "TLSv1.2"))) {
+            assertSslCommunicationMetrics(reg, 1, 1, 0);
+        }
+        assertSslCommunicationMetrics(reg, 1, 0, 0);
```
Comment (NSAmelchev) [inline modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:128]: Add line break, please

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            GridClient ignored = start(gridClientConfiguration("connectorClient", "trustthree", CIPHER_SUITE, "TLSv1.2"))
+        ) {
+            assertSslCommunicationMetrics(reg, 1, 1, 0);
+        }
+        assertSslCommunicationMetrics(reg, 1, 0, 0);
+
+        // Tests untrusted certificate.
+        try (GridClient ignored = start(gridClientConfiguration("client", "trustthree", CIPHER_SUITE, "TLSv1.2"))) {
+            // GridClient makes 2 additional silent connection attempts if an SSL error occurs.
+        }
```
Comment (NSAmelchev) [inline modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:163]: Add line break, please

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    }
+
+    /** Tests SSL metrics produced by thin client connection. */
+    @Test
+    public void testClientConnector() throws Exception {
+        MetricRegistry reg = mreg(startClusterNode(0), CLIENT_CONNECTOR_METRIC_REGISTRY_NAME);
+
+        try (IgniteClient ignored = startClient(clientConfiguration("thinClient", "trusttwo", CIPHER_SUITE, "TLSv1.2"))) {
+            assertSslCommunicationMetrics(reg, 1, 1, 0);
+        }
```
Comment (NSAmelchev) [inline modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:235]: Add line break, please

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerProcessor.java [hunk 5]
```diff
@@ -96,6 +96,9 @@
     /** Executor service. */
     private ExecutorService execSvc;
 
+    /** The name of the metric registry associated with the thin client connector. */
+    public static final String CLIENT_CONNECTOR_METRIC_REGISTRY_NAME = metricName("client", "connector", "tcp");
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerProcessor.java:100]: Can we move this to class constants?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java [hunk 6]
```diff
@@ -201,6 +231,20 @@ public void enabledProtocols(String... enabledProtos) {
 
             sslMeta.handler(hnd);
 
+            if (handshakeDurationHistogram != null) {
+                GridNioFutureImpl<?> fut = ses.meta(HANDSHAKE_FUT_META_KEY);
+
+                if (fut == null) {
+                    fut = new GridNioFutureImpl<>(null);
+
+                   ses.addMeta(HANDSHAKE_FUT_META_KEY, fut);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java:240]: Fix tabulation, please

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java [hunk 7]
```diff
@@ -72,6 +76,18 @@
     /** Whether direct mode is used. */
     private boolean directMode;
 
+    /** The name of the metric that provides histogram of SSL handshake duration. */
+    public static final String SSL_HANDSHAKE_DURATION_HISTOGRAM_METRIC_NAME = "sslHandshakeDurationHistogram";
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java:80]: Most metric names start with a capital letter. Should we rename these metrics too?

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java (hunk 2 above)
Comment (ololo3000) [reply modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:129]: Done.

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java (hunk 3 above)
Comment (ololo3000) [reply modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:168]: Done.

File: modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java (hunk 4 above)
Comment (ololo3000) [reply modules/clients/src/test/java/org/apache/ignite/common/SslConnectorsMetricTest.java:243]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerProcessor.java (hunk 5 above)
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/ClientListenerProcessor.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java (hunk 7 above)
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java (hunk 6 above)
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/ssl/GridNioSslFilter.java]: Done.

**Review by NSAmelchev — APPROVED** [review]

**Pull request merged**